### ISSUE TITLE
Exclude inactive components from model building entirely.

### ISSUE THREAD

**Issue by koen-vg** [PyPSA/PyPSA#1056]
I'm liking the new `active` attribute, but the next step should hopefully be to exclude inactive components entirely from model building and optimisation. The impacts on memory consumption are substantial.

I've implemented a rough version of the idea in https://github.com/PyPSA/PyPSA/commit/094af75bf5d2f31ccfca39b8b4644e32a25bfb53. It was honestly a little messy, and I wonder if there's a better way of doing things. In the end, there just hasn't been the assumption throughout the code that components might not be enabled in optimisations, so there are a few places where the `active` attribute has to be dealt with more actively.

It was mentioned in https://github.com/PyPSA/PyPSA/pull/1038 that the feature I'm suggesting should be on the horizon, but possibly only after https://github.com/PyPSA/PyPSA/issues/848. My implementation didn't particularly need the latter to be implemented, but maybe it will make things easier in a way I couldn't see?

Excluding inactive components from model-building is necessary in order to get the benefit in terms of memory of build year aggregation (see https://github.com/PyPSA/pypsa-eur/pull/1056).

I will also add, however, that in testing https://github.com/PyPSA/pypsa-eur/pull/1056, I've also found that custom constraints in pypsa-eur in particular will also need to be updated to take `active` into account if this feature is implemented.

**Comment by euronion** [comment]
I haven't been following any of this at all. That being said let me drop an idea that might be very simple to implement:

At the entry point for building the model before the optimisation, create a copy of the network and remove all `active=False` components from it. Then do everything as it was until now.
After the optimisation, update the original network with the optimised network (like a `pandas.DataFrame.update(...)`), thus leaving intact the `active=False` components.

**Comment by koen-vg** [comment]
That's certainly a possibility! I would come with some memory overhead, but not all too much in the grand scheme of things (and certainly much less than the overhead of creating (unused) optimisation variables for inactive components).

Maybe what should be considered explicitly before deciding which way to go with the implementation is the following: should the `active` attribute play any role in the output of various utility functions such as `get_switchable_as_dense`, `get_active_assets`, etc.? (See also the discussion in https://github.com/PyPSA/PyPSA/pull/1038.) To me, it would make sense at the very least if `get_active_assets` respected the `active` attribute, but I would also extend that to `get_extendable_i` for example.

Downstream projects defining custom constraints (such as pypsa-eur) will have to adapt one way or the other if they decide to use `active` (which pypsa-eur should; see my PR about build year aggregation). It's always possible to manually filter out inactive components when defining constraints, but it might be easier/cleaner to rely on `get_extendable_i`, `get_active_assets` if they take care of filtering out inactive components?

I'm sure you have some thoughts on this too, @FabianHofmann

**Comment by lkstrp** [comment]
I completely agree, this can be implemented much better. I will work on moving the whole component logic into a class soon, to remove all these arbitrary nested calls and have a unified interface to access any component data. And also close #848. This would allow for a much cleaner, but same approach as in https://github.com/PyPSA/PyPSA/commit/094af75bf5d2f31ccfca39b8b4644e32a25bfb53. 

Just a small sketch, but instead of
```python
  static = n.static(component)
  dynamic = n.dynamic(component)

  index = static.loc[n.get_active_assets(component)].index

  varying_i = dynamic[attr].columns
  fixed_i = static.index.difference(varying_i)
```
we could have something like (just changing the api)
```python
  c = n.as_component(c) # component class, passed argument can be arbitrary

  index = c.static_active.index 

  varying_i = c.dynamic[attr].columns
  fixed_i = c.static.index.difference(varying_i)
```
and introduce next do `c.dynamic` and `c.static` an view which filters for active. Not sure if this should be `c.static_active` or even 

```python
c.static.select(active_assets=True)
``` 
as a custom method on top of the DataFrame. This would also allow to define other custom filter methods which then could be easily applied and would be much cleaner. But the implementation of both is very similar. Just some ideas.

While being a good hot fix, the filtered down network copy just brings more chaos and doesn't allow for granular changes. So if this has time, I would wait until the changes above are implemented.

> To me, it would make sense at the very least if `get_active_assets` respected the `active` attribute, but I would also extend that to `get_extendable_i` for example.

`get_active_assets` respects the `active` attribute. But I agree, this needs to be nailed down for all utility functions. Which makes me like this generalised way of filtering on components even more.

**Comment by lkstrp** [comment]
Do we wanna put this in `1.0` @fneum @FabianHofmann ? Now with `new-opt` close to being merged

**Comment by FabianHofmann** [comment]
yes! let's do that. I like the idea of having a selection on the component object. if we want to make our lifes easy we could add patterns like this to variables.py/constraints.py 

```
    c = n.components[c_name]
    com_i = c.committables.intersection(c.actives)
```

**Comment by fneum** [comment]
Yes! Let's scope how easy this is to do.
